dmoj/judge.py

```diff
@@ -459,69 +491,125 @@ def _grade_cases(self) -> Generator[Tuple[IPC, tuple], None, None]:
-                case_number += 1
-
-                # Stop grading if we're short circuiting
-                if is_short_circuiting:
-                    result = Result(case, result_flag=Result.SC)
+        global_test_number = 0
+
+        class GradingAbort(Exception):
+            pass
+
+        def grade(case) -> Result:
+            result = self.grader.grade(case)
+
+            # If the submission was killed due to a user-initiated abort, any result is meaningless.
+            if self._abort_requested:
+                raise GradingAbort
+
+            # Legacy hack: we need to allow graders to read and write `proc_output` on the `Result` object, but the
+            # judge controller only cares about the trimmed output, and shouldn't waste memory buffering the full
+            # output. So, we trim it here so we don't run out of memory in the controller.
+            result.proc_output = result.output
+            return result
+
+        def skip_single(case) -> Tuple[IPC, tuple]:
+            nonlocal batch_number, global_test_number
+            global_test_number += 1
+            return IPC.RESULT, (batch_number, global_test_number, Result(case, result_flag=Result.SC))
+
+        def skip_batched_cases(batched_cases: Iterable[BatchedTestCase]) -> Generator[Tuple[IPC, tuple], None, None]:
+            nonlocal batch_number, global_test_number
+            for case in batched_cases:
+                yield skip_single(case)
+
+        def grade_batched_cases(batched_cases: List[BatchedTestCase]) -> Generator[Tuple[IPC, tuple], None, int]:
+            nonlocal batch_number, global_test_number
+            case_iter = iter(batched_cases)
+            for case in case_iter:
+                result = grade(case)
+
+                global_test_number += 1
+                yield IPC.RESULT, (batch_number, global_test_number, result)
+
+                if result.result_flag & Result.WA:
+                    # Skip the rest.
+                    yield from skip_batched_cases(case_iter)
+                    return Result.WA
+
+            return Result.AC
+
+        def grade_many(
+            tests: List[TopLevelCase], should_run: Callable[[TopLevelCase, Set[int]], bool]
+        ) -> Generator[Tuple[IPC, tuple], None, Set[int]]:
+            nonlocal batch_number, global_test_number, skip_all
+            failed_tests: Set[int] = set()
+            for test in tests:
+                failed = False
+                skip_current = skip_all or not should_run(test, failed_tests)
+                if isinstance(test, Batch):
+                    batch_number += 1
+                    yield IPC.BATCH_BEGIN, (batch_number,)
+
+                    if skip_current:
+                        yield from skip_batched_cases(test.batched_cases)
+                    else:
+                        batch_verdict = yield from grade_batched_cases(test.batched_cases)
+                        failed = batch_verdict == Result.WA
```

Comment:
Note to reviewer: we can't condense this AFAICT, it's actually a syntax error if we try.